### DISCUSSION TITLE
Licensing issue with openssl <3.0

### DISCUSSION THREAD

**Discussion by dionysius** [dani-garcia/vaultwarden#3067] [Q&A]
While building my debian packaging I got informed by the tooling that vaultwarden, [licensed as GPL-3](https://github.com/dani-garcia/vaultwarden/blob/main/LICENSE.txt), links to openssl, [licensed as either Apache 2.0 (v3.0+) or the dual OpenSSL and SSLeay license (<v3.0)](https://www.openssl.org/source/license.html), which can be incompatible.  _Apache 2.0 seems fine from my mind, **but I'm not fit in that licensing jungle**_. The OpenSSL license for <3.0 seem to be the issue.

[Linking a stackoverflow question which summerizes the situation and solution pretty well.](https://opensource.stackexchange.com/questions/2233/gpl-v3-with-openssl-exception)

Current Ubuntu LTS jammy uses openssl 3.0.0, but current Debian stable bullseye uses openssl 1.1.1. It's good that we soon don't have to worry anymore once openssl 3.0.0 is rolled out everywhere (implicates that my above assumption is correct regarding Apache 2.0).

Is it possible to include an exception for openssl as they've discussed in the above stackoverflow question? So far I couldn't find such an exception in the current license text yet.

**Comment by stefan0xC** [comment]
Does the system library exception not apply? As far as I've looked into this, Debian apparently [decided to treat OpenSSL as a system library](https://bugs.debian.org/972181) like Fedora but I'm not a lawyer and not sure if an exception is still necessary or not, and to what extend the usage of `openssl` crate (which could theoretically be also linked against LibreSSL) makes a difference in this case or not.

Since the Apache 2.0 license [is compatible](https://www.apache.org/licenses/GPL-compatibility.html) you could ship a vendored `openssl-3.0` instead of linking to the system library as a workaround (if Debian packaging policies allow this).

**Comment by BlackDex** [comment]
I think Debian/Ubuntu isn't an issue atm. But the Alpine based images which are using 1.1.1 as a static library might be. Switching to 3.x is maybe something we should think about. Also with the debian images. But i rather wait until bookworm is stable.

**Comment by dionysius** [comment]
I worked a bit more on my packaging and ended up not shipping a build for debian anyway since github actions has no debian images. BTW its the debian tooling that informed me, so I'd have to dig deeper why I still get the warning if debian would have such exception globally.

For me it would be fine closing this. I'm right now not affected by this anymore and the issue will anyway silently go away in time. I just also wanted to let you know.

**Comment by stefan0xC** [comment]
@BlackDex Ah, you're right that this applies to the releases of Vaultwarden as well but I think then also the Debian based images are affected (since they use or atleast are linked against `libssl.so.1.1`).

If I understand the license issue correctly then granting an additional permission for linking to OpenSSL 1.1 [as suggested by the GNU project](https://www.gnu.org/licenses/gpl-faq.en.html#GPLIncompatibleLibs) and requested here would allow ourselves (and downstream) to also link against openssl-1.1 (especially in the case of the static builds where a system library exception is not applicable) because [the license](https://www.openssl.org/source/license-openssl-ssleay.txt) of OpenSSL 1.1 has an advertisement clause and would otherwise be [incompatible with GPLv3](https://www.gnu.org/licenses/license-list.en.html#OpenSSL). 

But even then, we'd still need to abide by the additional license terms of its license in those cases, so I'm not sure if this is something we want or need. So I think we should probably switch to openssl-3.x in all Docker images as soon as possible to be on the safe side.

**Reply by BlackDex** [reply]
We can only switch when Debian bookworm is stable. That has OpenSSL v3
I can see if i can already switch the MUSL build images though.
But i actually would like to keep that as close as possible to the Debian library versions.
Looks like i also need to try and build PostgreSQL 15, since that is also used in Bookworm.
